dhis-2/dhis-api/src/main/java/org/hisp/dhis/util/StreamUtils.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright (c) 2004-2023, University of Oslo
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ * Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ *
+ * Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ * Neither the name of the HISP project nor the names of its contributors may
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.util;
+
+import java.util.Collection;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+/**
+ * @author David Mackessy
+ */
+public class StreamUtils
+{
+
+    private StreamUtils()
+    {
+        throw new IllegalStateException( "StreamUtils is a utility class and cannot be instantiated" );
+    }
+
+    /**
+     * This method allows calling stream on a collection, whether it is null or
+     * non-null, returning and empty stream when passed null as an argument or
+     * the actual stream of elements if non-null. This will help reduce
+     * NullPointExceptions.
+     *
+     * @param collection of any type
+     * @return a stream of elements (of Type T) from the collection if the
+     *         collection is not null or, return an empty stream if the
+     *         collection is null.
+     * @param <T> is the Type of element
+     */
+    public static <T> Stream<T> nullSafeCollectionToStream( Collection<T> collection )
```

Comment:
`streamOf`` with nullness annotations should do it and result in more readable code. Not sure if someone mentioned for you that we have the convention that generally we do assume any collection type to be nun-null. If you have code that makes different assumptions it should definitely use the nullness annotation.

Reply:
I can rename this. I haven't heard that convention from anyone yet about assuming any collection is not null. That's a very big assumption but happy to go along with what we all do. I have seen that we usually instantiate an empty collection when declaring it as a class property, is that the main/only way we base our assumption for this?

Reply:
This differs from case to case. Also there is much code not up to speed with this. Just remember this for any changes or new code to code so that collections are never null unless it is a special case e.g. results from third party code and code that handles such cases to adapt to our assumptions. We also have the convention to assume that any collection received may be immutable. Again this is not true for private methods that know their caller but a public method receiving a collection should assume it cannot be changed.